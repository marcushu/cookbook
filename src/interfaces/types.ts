export interface Recipe {
    name: string
    owner: string
    description: string
    ingredients: string[]
    instructions: string
    breakfast: boolean
    lunch: boolean
    dinner: boolean
    glutenFree: boolean
    vegan: boolean
    vegetarian: boolean
    imageUrl: string
  }

  export interface ShoppinglistIngredient {
    ingredient: string
    recipe: string // optional to allow a user to add ingredients w/out a recipe
    /** used if the user wants the ingredients listed in a particular order */
    order?: number
  }

  export type IngredientSortPolicy = 'recipe' | 'sorting'

  export interface UserState {
    userName: string
    orderIngredientsBy: IngredientSortPolicy
    /** recipes created by this user. These exist in a different db collection
     * and need to be loaded separately. */
    recipes: Recipe[]
    favorites: Recipe[]
    shoppingList: ShoppinglistIngredient[]
  }