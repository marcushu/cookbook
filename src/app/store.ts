import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import userReducer from '../redux/userState';
import searchResultsReducer from '../redux/searchResultState';

export const store = configureStore({
  reducer: {
    user: userReducer,
    searchResults: searchResultsReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
