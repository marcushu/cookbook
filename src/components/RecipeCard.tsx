import { FunctionComponent, useState } from "react";
import { addFavorite, selectUserName } from "../redux/userState";
import { Recipe } from "../interfaces/types";
import { useAppDispatch, useAppSelector } from "../app/hooks";
import { Box, Button, Grid, IconButton, Paper, styled, Tooltip, Typography } from "@mui/material";
import FavoriteIcon from '@mui/icons-material/Favorite';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { grey } from "@mui/material/colors";
import FullRecipe from "./FullRecipe";
import RestrictionTags from "./RestrictionTags";
import MealtimeTags from "./MealtimeTags";

interface RecipeCardProps {
  recipe: Recipe
}

const RecipePaper = styled(Paper)({
  margin: '8px auto 8px auto',
  width: '900px'
});

const RecipeFooter = styled('div')({
  borderTop: 'solid',
  borderColor: grey[400],
  borderWidth: '1px',
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  marginLeft: '15px'
});

const ImageGridItem = styled(Grid)({
  width: '300px',
  height: '300px',
  overflow: 'hidden',
  textAlign: 'center'
});

const TextGridItem = styled(Grid)({
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'space-between'
})

const RecipeCard: FunctionComponent<RecipeCardProps> = ({ recipe }) => {
  const { name, imageUrl, owner, description } = recipe;
  const userName = useAppSelector(selectUserName);
  const [showFullRecipe, setShowFullRecipe] = useState(false);
  const dispatch = useAppDispatch();
  
  const addToFavorites = () => {
    dispatch(addFavorite({recipe, userName}));
  }

  return (
    <RecipePaper variant='outlined'>
      <Grid container>
        <ImageGridItem item xs={12} sm={4} >
          <img style={{ maxWidth: '100%' }} src={imageUrl} alt="food" />
        </ImageGridItem>
        <TextGridItem item xs={12} sm={8}>
          <Box component='div' pl={2}>
            <Typography pt={2} display='flex' justifyContent='space-between' variant='h4'>
              {name}
              {userName &&
                <Tooltip title='Add to favorites'>
                  <IconButton onClick={addToFavorites}>
                    <FavoriteIcon color='primary' />
                  </IconButton>
                </Tooltip>}
            </Typography>
            <Typography variant='subtitle2'>{owner}</Typography>
            <Typography pt={2}>{description}</Typography>
            <MealtimeTags recipe={recipe} />
          </Box>
          <RecipeFooter>
            <RestrictionTags recipe={recipe} />
            <Box component='div'>
              {!showFullRecipe &&
                <Button
                  onClick={() => setShowFullRecipe(!showFullRecipe)}
                  startIcon={<ExpandMoreIcon fontSize='large' />} >
                  more
                </Button>}
            </Box>
          </RecipeFooter>
        </TextGridItem>
        {showFullRecipe &&
          <FullRecipe
            hideMe={() => setShowFullRecipe(!showFullRecipe)}
            recipe={recipe} />}
      </Grid>
    </RecipePaper >
  );
}

export default RecipeCard;