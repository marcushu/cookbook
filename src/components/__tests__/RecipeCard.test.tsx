import { render } from "@testing-library/react"
import { Recipe } from "../../interfaces/types";
import { Provider } from 'react-redux';
import RecipeCard from '../RecipeCard';
import { store } from '../../app/store';

describe('RecipeCard', () => {
    const mockRecipe = {
        dinner: true,
        lunch: false,
        name: 'mock name',
        owner: 'mock owner',
        glutenFree: true,
        ingredients: ['ing1', 'ing2'],
        vegan: true,
        description: 'mock description',
        instructions: 'these are the mock instructions',
        imageUrl: 'http://mock.url'
    } as Recipe;

    it('renders', () => {
        const { asFragment } = render(<Provider store={store}><RecipeCard recipe={mockRecipe} /></Provider>)

        expect(asFragment()).toMatchSnapshot();
    });

    it('displays meal type', () => {
        const { getByText } = render(<Provider store={store}><RecipeCard recipe={mockRecipe} /></Provider>);

        const mealTypeDinner = getByText(/Dinner/i);

        expect(mealTypeDinner).toBeDefined();
    });

    it('displays meal restriction types', () => {
        const { getByText } = render(<Provider store={store}><RecipeCard recipe={mockRecipe} /></Provider>);

        const restriction = getByText(/Vegan/i);
        const restriction2 = getByText(/GF/i);

        expect(restriction).toBeDefined();
        expect(restriction2).toBeDefined();
    });

    it('displays recipe and description', () => {
        const { getByText } = render(<Provider store={store}><RecipeCard recipe={mockRecipe} /></Provider>);

        const recipeName = getByText(/mock name/i);
        const description = getByText(/mock description/i);

        expect(recipeName).toBeDefined();
        expect(description).toBeDefined();
    });
})