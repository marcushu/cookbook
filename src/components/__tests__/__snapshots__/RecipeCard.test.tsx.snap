// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeCard renders 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-1u15cpw-MuiPaper-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-n7zmlg-MuiGrid-root"
      >
        <img
          alt="food"
          src="http://mock.url"
          style="width: 100%;"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 css-gzpq3j-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-n085mf"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-owqfbr-MuiTypography-root"
          >
            mock name
          </h4>
          <div
            class="MuiBox-root css-a8cz4w"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
            >
              moc*******
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1hd3sn1-MuiTypography-root"
            >
              mock description
            </p>
            <div
              class="MuiBox-root css-1sf3xto"
            >
              <div
                class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-v7lgux-MuiChip-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconMedium MuiChip-iconColorDefault css-i4bv87-MuiSvgIcon-root"
                  data-testid="CheckIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M9 16.17 4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
                  />
                </svg>
                <span
                  class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
                >
                  Dinner
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-1r0oefp"
        >
          <div
            class="MuiBox-root css-1jpm9pd"
          >
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-filledSecondary css-15ma645-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
              >
                GF
              </span>
            </div>
            <div
              class="MuiChip-root MuiChip-filled MuiChip-sizeSmall MuiChip-colorSecondary MuiChip-filledSecondary css-15ma645-MuiChip-root"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall css-wjsjww-MuiChip-label"
              >
                vegan
              </span>
            </div>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              more
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-11uq76c-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-1nb6n7u-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 css-14aae1h-MuiGrid-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                >
                  <li>
                    ing1
                  </li>
                </div>
                <div
                  class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                >
                  <li>
                    ing2
                  </li>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-8 css-oq1gac-MuiGrid-root"
              >
                <div
                  class="MuiBox-root css-t0051x"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-p3gdy2-MuiTypography-root"
                  >
                    ~  these are the mock instructions 
                    <br />
                     
                    <br />
                  </p>
                </div>
                <div
                  class="MuiBox-root css-1fuqp3k"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
                        data-testid="ExpandLessIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
                        />
                      </svg>
                    </span>
                    less
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
